### DISCUSSION TITLE
How to Start code-server installed from npm?

### DISCUSSION THREAD

**Discussion by Slodziak190q** [coder/code-server#4154] [Q&A] [answered]
J installed code-server on rpi (again) beacouse railway was to laggy but j cant start it 
When j type code-server --link in terminal i get an error

**Comment by jsjoeio** [comment]
If it's installed with npm, double check by running `code-server -v` and it should print the version.

To run, run `code-server`. 

To run with link, run `code-server --link`. 

You may have to restart your terminal. Let us know what happens.

**Comment by lg-nineoneninethreenine** [comment]
Can you show a screenshot of the error?

**Reply by Slodziak190q** [reply]
![Screenshot_20210914_153043_com minisatip putty](https://user-images.githubusercontent.com/87336535/133266658-63698c08-a6da-4d28-be2c-73d168f2421f.jpg)

**Reply by lg-nineoneninethreenine** [reply]
What is that? It's a Raspberry Pi, right?

**Reply by Slodziak190q** [reply]
Yes its my raspbbery pi but j ssh to it

**Comment by lg-nineoneninethreenine** [comment, marked as answer]
I am a fool, I thought the installation was wrong, I think he ran `npm install code-server` _instead_ of `npm install -g code-server` which enables code-server globally. So try that and report bach here.

If you see the screenshot(ya I noticed he used Android for the Pi connection), it doesn't find code-server. So it is not enabled globally.

Screenshot for ref:
> ![Screenshot_20210914_153043_com minisatip putty](https://user-images.githubusercontent.com/87336535/133266658-63698c08-a6da-4d28-be2c-73d168f2421f.jpg)
> 

Originally posted by @Slodziak190q on my thread.

**Reply by Slodziak190q** [reply]
ok thanks j will try npm install -g code-server

**Reply by lg-nineoneninethreenine** [reply]
Yeah @Slodziak190q one small question: how do you get network? Via Ethernet port?

**Reply by Slodziak190q** [reply]
j dont understand j ssh to my raspbbery pi with port 22

**Reply by lg-nineoneninethreenine** [reply]
How do you connect network to your Raspberry Pi?

**Reply by Slodziak190q** [reply]
with ethernet cable (sorry jm from poland my english is bad)

**Reply by Slodziak190q** [reply]
Got another problem 
![sendgithubXd](https://user-images.githubusercontent.com/87336535/133271403-eab699bf-0892-4438-9af6-0678d4e7cb08.jpg)

**Reply by lg-nineoneninethreenine** [reply]
Can you share the logfile here?

**Reply by Slodziak190q** [reply]
0 info it worked if it ends with ok
1 verbose cli [
1 verbose cli   '/usr/bin/node',
1 verbose cli   '/usr/bin/npm',
1 verbose cli   'install',
1 verbose cli   '-global',
1 verbose cli   'code-server'
1 verbose cli ]
2 info using npm@6.14.13
3 info using node@v14.17.0
4 verbose npm-session a81636aaa8283fed
5 silly install loadCurrentTree
6 silly install readGlobalPackageData
7 http fetch GET 200 https://registry.npmjs.org/code-server 61ms (from cache)
8 silly pacote tag manifest for code-server@latest fetched in 107ms
9 timing stage:loadCurrentTree Completed in 197ms
10 silly install loadIdealTree
11 silly install cloneCurrentTreeToIdealTree
12 timing stage:loadIdealTree:cloneCurrentTree Completed in 1ms
13 silly install loadShrinkwrap
14 timing stage:loadIdealTree:loadShrinkwrap Completed in 5ms
15 silly install loadAllDepsIntoIdealTree
16 silly resolveWithNewModule code-server@3.11.1 checking installable status
17 http fetch GET 200 https://registry.npmjs.org/@coder%2flogger 188ms (from cache)
18 http fetch GET 200 https://registry.npmjs.org/cookie-parser 164ms (from cache)
19 http fetch GET 200 https://registry.npmjs.org/argon2 189ms (from cache)
20 http fetch GET 200 https://registry.npmjs.org/http-proxy 162ms (from cache)
21 http fetch GET 200 https://registry.npmjs.org/httpolyglot 164ms (from cache)
22 http fetch GET 200 https://registry.npmjs.org/js-yaml 170ms (from cache)
23 silly pacote version manifest for @coder/logger@1.1.16 fetched in 218ms
24 silly resolveWithNewModule @coder/logger@1.1.16 checking installable status
25 silly pacote range manifest for argon2@^0.28.0 fetched in 217ms
26 silly resolveWithNewModule argon2@0.28.2 checking installable status
27 silly pacote range manifest for cookie-parser@^1.4.5 fetched in 198ms
28 silly resolveWithNewModule cookie-parser@1.4.5 checking installable status
29 silly pacote range manifest for http-proxy@^1.18.0 fetched in 193ms
30 silly resolveWithNewModule http-proxy@1.18.1 checking installable status
31 silly pacote range manifest for httpolyglot@^0.1.2 fetched in 193ms
32 silly resolveWithNewModule httpolyglot@0.1.2 checking installable status
33 silly pacote range manifest for js-yaml@^4.0.0 fetched in 215ms
34 silly resolveWithNewModule js-yaml@4.1.0 checking installable status
35 http fetch GET 200 https://registry.npmjs.org/proxy-agent 69ms (from cache)
36 http fetch GET 200 https://registry.npmjs.org/proxy-from-env 70ms (from cache)
37 http fetch GET 200 https://registry.npmjs.org/pem 76ms (from cache)
38 http fetch GET 200 https://registry.npmjs.org/rotating-file-stream 72ms (from cache)
here you go Idk if its good j cant access more

**Reply by jsjoeio** [reply]
This looks like an issue with Argon2. I think you need to make sure your environment has some global dependencies installed. 

> You MUST have a node-gyp global install before proceeding with install, along with GCC >= 5 / Clang >= 3.3. On Windows, you must compile under Visual Studio 2015 or newer.
> 
> node-argon2 works only and is tested against Node >=12.0.0.

See [here](https://github.com/ranisalt/node-argon2#prebuilt-binaries)

**Reply by lg-nineoneninethreenine** [reply]
@jsjoeio do you have a Raspberry Pi? If you have, did you manage to see this the first time you tried getting code-server on it? Another doubt: can you install Argon and the other deps separately?

**Reply by lg-nineoneninethreenine** [reply]
Also @slodziak190q in the image it mentioned a path to the logfile right? Share the logfile from the Raspberry Pi or transfer the files via WinSCP or something and _upload_ the file here(drag drop into comment section) so that I can observe and find the major error.

**Reply by jsjoeio** [reply]
> @jsjoeio do you have a Raspberry Pi? If you have, did you manage to see this the first time you tried getting code-server on it? Another doubt: can you install Argon and the other deps separately?

Yes! I do. I installed it using `npm` but that was before we used Argon2. I'll need to do it again. My Pi isn't in an easy place to go test but I'll add it to my todo list and hopefully get to it next week!

**Reply by Slodziak190q** [reply]
> Also @Slodziak190q in the image it mentioned a path to the logfile right? Share the logfile from the Raspberry Pi or transwer the files via WinSCP or something and _upload_ the file here(drag drop into comment section) so that I can observe and find the major error.

but the problem is j can't access the file it says j need to be root but from terminal j can only do nano (log file name)

**Reply by lg-nineoneninethreenine** [reply]
I think you _can_ access the entire logfile. By default, the Pi's default user is "pi". So, what I think is that you `sudo su` into root, and type this:
```bash
cp -r /root/.npm/_logs/2021-09-14T13_56_34_512Z-debug.log /home/pi/Documents/2021-09-14T13_56_34_512Z-debug.log
```
Then, you turn your computer on(hope you have Windows or macOS), get WinSCP(it can share files easily), go to Google and search "how to access raspberry pi via winscp?". Then, learn how to access(idk one bit) your Pi via WinSCP. After you access, go to Documents folder and drag-drop the log file here. That can make it easy since it might not be just Argon2.

**Reply by lg-nineoneninethreenine** [reply]
And yeah, the command is big since the path and the filename is big.

**Reply by Slodziak190q** [reply]
ok j downloaded it form pi here you go
[2021-09-14T13_56_34_512Z-debug.log](https://github.com/cdr/code-server/files/7177872/2021-09-14T13_56_34_512Z-debug.log)

**Reply by Slodziak190q** [reply]
> This looks like an issue with Argon2. I think you need to make sure your environment has some global dependencies installed.
> 
> > You MUST have a node-gyp global install before proceeding with install, along with GCC >= 5 / Clang >= 3.3. On Windows, you must compile under Visual Studio 2015 or newer.
> > node-argon2 works only and is tested against Node >=12.0.0.
> 
> See [here](https://github.com/ranisalt/node-argon2#prebuilt-binaries)

J cant install this homebrew (it says j must have intel procesor its not possible on pi) thing any other solutions how to install argon?

**Reply by lg-nineoneninethreenine** [reply]
@jsjoeio and @Slodziak190q I finally managed to recreate. Looks like the Pi I ordered came today so I tried installing and I got the error. It _is_ Argon2 as @jsjoeio told. Looks like it's a bug.

**Reply by jsjoeio** [reply]
Dang! @im-coder-lg thank you for confirming. Any ideas come to mind to address this? Otherwise, we can open an issue and add to backlog to fix later!

**Reply by lg-nineoneninethreenine** [reply]
I couldn't install Argon2 separately and even got node-pre-gyp but it doesn't work. I have to check about that tomorrow.

**Reply by jsjoeio** [reply]
No worries at all. Really really appreciate your help <3

**Reply by lg-nineoneninethreenine** [reply]
@jsjoeio I think I need to test the same on my WSL. It could be the fact that since a Pi is x86 system architecture this could be happening.

**Reply by lg-nineoneninethreenine** [reply]
@jsjoeio I found something. Looks like as you told it is Argon2 but it affects the npm package globally. Take these logs for example:
```console
> yarn@1.22.11 preinstall /usr/lib/node_modules/code-server/node_modules/yarn
> :; (node ./preinstall.js > /dev/null 2>&1 || true)

/usr/bin/code-server -> /usr/lib/node_modules/code-server/out/node/entry.js

> argon2@0.28.2 install /usr/lib/node_modules/code-server/node_modules/argon2
> node-pre-gyp install --fallback-to-build

node-pre-gyp WARN Pre-built binaries not installable for argon2@0.28.2 and node@14.17.6 (node-v83 ABI, glibc) (falling back to source compile with node-gyp) 
node-pre-gyp WARN Hit error EACCES: permission denied, mkdir '/usr/lib/node_modules/code-server/node_modules/argon2/lib' 
gyp WARN EACCES current user ("nobody") does not have permission to access the dev dir "/root/.cache/node-gyp/14.17.6"
gyp WARN EACCES attempting to reinstall using temporary dev dir "/usr/lib/node_modules/code-server/node_modules/argon2/.node-gyp"
gyp WARN install got an error, rolling back install
gyp WARN install got an error, rolling back install
gyp ERR! configure error 
gyp ERR! stack Error: EACCES: permission denied, mkdir '/usr/lib/node_modules/code-server/node_modules/argon2/.node-gyp'
gyp ERR! System Linux 5.4.0-1046-gke
gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "configure" "--fallback-to-build" "--module=/usr/lib/node_modules/code-server/node_modules/argon2/lib/binding/napi-v3/argon2.node" "--module_name=argon2" "--module_path=/usr/lib/node_modules/code-server/node_modules/argon2/lib/binding/napi-v3" "--napi_version=8" "--node_abi_napi=napi" "--napi_build_version=3" "--node_napi_label=napi-v3"
gyp ERR! cwd /usr/lib/node_modules/code-server/node_modules/argon2
gyp ERR! node -v v14.17.6
gyp ERR! node-gyp -v v5.1.0
gyp ERR! not ok 
node-pre-gyp ERR! build error 
node-pre-gyp ERR! stack Error: Failed to execute '/usr/bin/node /usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --module=/usr/lib/node_modules/code-server/node_modules/argon2/lib/binding/napi-v3/argon2.node --module_name=argon2 --module_path=/usr/lib/node_modules/code-server/node_modules/argon2/lib/binding/napi-v3 --napi_version=8 --node_abi_napi=napi --napi_build_version=3 --node_napi_label=napi-v3' (1)
node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/code-server/node_modules/@mapbox/node-pre-gyp/lib/util/compile.js:89:23)
node-pre-gyp ERR! stack     at ChildProcess.emit (events.js:400:28)
node-pre-gyp ERR! stack     at maybeClose (internal/child_process.js:1055:16)
node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)
node-pre-gyp ERR! System Linux 5.4.0-1046-gke
node-pre-gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/code-server/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
node-pre-gyp ERR! cwd /usr/lib/node_modules/code-server/node_modules/argon2
node-pre-gyp ERR! node -v v14.17.6
node-pre-gyp ERR! node-pre-gyp -v v1.0.5
node-pre-gyp ERR! not ok 
Failed to execute '/usr/bin/node /usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --module=/usr/lib/node_modules/code-server/node_modules/argon2/lib/binding/napi-v3/argon2.node --module_name=argon2 --module_path=/usr/lib/node_modules/code-server/node_modules/argon2/lib/binding/napi-v3 --napi_version=8 --node_abi_napi=napi --napi_build_version=3 --node_napi_label=napi-v3' (1)
npm WARN @coder/logger@1.1.16 requires a peer of @google-cloud/logging@^4.5.2 but none is installed. You must install peer dependencies yourself.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! argon2@0.28.2 install: `node-pre-gyp install --fallback-to-build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the argon2@0.28.2 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-09-17T05_24_00_262Z-debug.lo
```
This was done on a Gitpod workspace since my WSL has errors. 

Neofetch results on Gitpod:
```console
            .-/+oossssoo+/-.
        `:+ssssssssssssssssss+:`
      -+ssssssssssssssssssyyssss+-
    .ossssssssssssssssssdMMMNysssso.
   /ssssssssssshdmmNNmmyNMMMMhssssss/
  +ssssssssshmydMMMMMMMNddddyssssssss+
 /sssssssshNMMMyhhyyyyhmNMMMNhssssssss/
.ssssssssdMMMNhsssssssssshNMMMdssssssss.
+sssshhhyNMMNyssssssssssssyNMMMysssssss+
ossyNMMMNyMMhsssssssssssssshmmmhssssssso
ossyNMMMNyMMhsssssssssssssshmmmhssssssso
+sssshhhyNMMNyssssssssssssyNMMMysssssss+   gitpod@ws-96aebd8e-81c6-4fdc-9215-6d34d01aedf1 
.ssssssssdMMMNhsssssssssshNMMMdssssssss.   ---------------------------------------------- 
 /sssssssshNMMMyhhyyyyhdNMMMNhssssssss/    OS: Ubuntu 20.04.2 LTS x86_64 
  +sssssssssdmydMMMMMMMMddddyssssssss+     Host: Google Compute Engine 
   /ssssssssssshdmNNNNmyNMMMMhssssss/      Kernel: 5.4.0-1046-gke 
    .ossssssssssssssssssdMMMNysssso.       Uptime: 6 days, 19 hours, 3 mins 
      -+sssssssssssssssssyyyssss+-         Packages: 665 (dpkg), 3 (brew) 
        `:+ssssssssssssssssss+:`           Shell: bash 5.0.17 
            .-/+oossssoo+/-.               Theme: Adwaita [GTK3] 
                                           Icons: Adwaita [GTK3] 
                                           Terminal: vscode 
                                           CPU: Intel Xeon (16) @ 2.800GHz 
                                           Memory: 46324MiB / 64319MiB
```
I need to get Neofetch on my RasPi. But it's Ubuntu 21.04. Looks like there is dependency issues.

**Reply by lg-nineoneninethreenine** [reply]
I tried locating Argon2 on package.json, look at this
https://github.com/cdr/code-server/blob/3c61d96d4fb63af50754027e032cac0bd32a0578/package.json#L81
and
https://github.com/cdr/code-server/blob/3c61d96d4fb63af50754027e032cac0bd32a0578/package.json#L86

Are they related to this? I think somebody will have to open a ticket on this bug. Only the npm package has bugs I guess.

**Reply by jsjoeio** [reply]
@im-coder-lg nice debugging 🙌 You're a beast. 

Okay, it's Friday morning and I'm feeling slow 😂 

Oh! Yes there was a `tar` vulnerability we had to patch. Are you saying that is the issue and that we should submit a patch upstream to `argon2`?

**Reply by lg-nineoneninethreenine** [reply]
Idk, I think something went wrong, noticed Docker also got some issue or something. I think this is a bug.

**Reply by lg-nineoneninethreenine** [reply]
But I do know this: everytime you try installing code-server from npm, you get an error of Argon2. Edit: Also while installing Argon2 I ended up with errors asking me to get node-pre-gyp. By the way, a general question, can I install Argon2 from npm's registry? If it's possible then I could try installing code-server after Argon2.

**Reply by lg-nineoneninethreenine** [reply]
> Oh! Yes there was a `tar` vulnerability we had to patch. Are you saying that is the issue and that we should submit a patch upstream to `argon2`?

Idk not much of a pro but if that's the issue then you have to check that whereas I will check Argon2. I have posted a link to this on the Slack channel for help there, need to check that.

**Reply by lg-nineoneninethreenine** [reply]
@jsjoeio Looks like I found Argon2 on the NPM registry [here](https://www.npmjs.com/package/argon2). Should I install it separately?

**Reply by Slodziak190q** [reply]
so how can j install code-server?

**Reply by lg-nineoneninethreenine** [reply]
I think you have to wait @Slodziak190q, looks like there is a bug. Try installing the old version.

**Reply by jsjoeio** [reply]
> @jsjoeio Looks like I found Argon2 on the NPM registry here. Should I install it separately?

That's the one we use in code-server! You shouldn't have to 🤔 But I guess you could try installing it globally

**Reply by Slodziak190q** [reply]
![image](https://user-images.githubusercontent.com/87336535/134190375-b961b3a6-4ad9-47f9-9916-474de4d378a1.png)
j tried to install old version of code-server but j get the error also as root j get terror
![image](https://user-images.githubusercontent.com/87336535/134191176-76a94781-01f5-4baa-b114-fe7a62252679.png)

**Reply by jsjoeio** [reply]
Ahh, I'm sorry. This has not been a fun process I'm sure. I need to dig into this more when I have time. In the meantime, I would suggest you deploy code-server following [our guides](https://github.com/cdr/deploy-code-server) (maybe Railway) that way you can have a smooth experience. 

We need to improve the Raspberry Pi experience.

**Comment by boyphongsakorn** [comment]
error timed out error write EPIPE after run code-server
i'm on raspberry pi zero w with `sudo npm install -g code-server --unsafe-perm=true --allow-root` command this work but can't run

**Reply by lg-nineoneninethreenine** [reply]
Try it as root.

**Comment by lg-nineoneninethreenine** [comment]
Thanks! This is what I needed!

**Reply by Slodziak190q** [reply]
and btw j have windows computer
